File: python3/leetcodepy/populating_next_right_pointers_in_each_node.py
"""
Given a binary tree

struct TreeLinkNode {
  TreeLinkNode *left;
  TreeLinkNode *right;
  TreeLinkNode *next;
}

Populate each next pointer to point to its next right node. If there is no next right node, the next pointer should be set to NULL.

Initially, all next pointers are set to NULL.

Note:

You may only use constant extra space.
Recursive approach is fine, implicit stack space does not count as extra space for this problem.
You may assume that it is a perfect binary tree (ie, all leaves are at the same level, and every parent has two children).

Example:

Given the following perfect binary tree,

     1
   /  \
  2    3
 / \  / \
4  5  6  7

After calling your function, the tree should look like:

     1 -> NULL
   /  \
  2 -> 3 -> NULL
 / \  / \
4->5->6->7 -> NULL
"""
from typing import Optional

from .utils import TreeLinkNode


class Solution1:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root: Optional[TreeLinkNode]):
        if root is None:
            return
        if root.left is not None:
            root.left.next = root.right
            if root.next is not None:
                root.right.next = root.next.left
            self.connect(root.left)
            self.connect(root.right)


class Solution2:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root: Optional[TreeLinkNode]):
        if root is None:
            return
        node = root
        dummy = TreeLinkNode(0)
        tail = dummy
        while node.left is not None:
            while node is not None:
                tail.next = node.left
                tail = tail.next
                tail.next = node.right
                tail = tail.next
                node = node.next
            node = dummy.next
            tail = dummy

File: java/src/main/java/io/binac/leetcode/PopulatingNextRightPointersInEachNode.java
package io.binac.leetcode;

/**
 * <p>You are given a <strong>perfect binary tree</strong> where all leaves are on the same level, and every parent has two children. The binary tree has the following definition:</p>
 *
 * <pre>struct Node {
 *   int val;
 *   Node *left;
 *   Node *right;
 *   Node *next;
 * }
 * </pre>
 *
 * <p>Populate each next pointer to point to its next right node. If there is no next right node, the next pointer should be set to <code>NULL</code>.</p>
 *
 * <p>Initially, all next pointers are set to <code>NULL</code>.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 * <img alt="" src="https://assets.leetcode.com/uploads/2019/02/14/116_sample.png" style="width: 500px; height: 171px;">
 * <pre><strong>Input:</strong> root = [1,2,3,4,5,6,7]
 * <strong>Output:</strong> [1,#,2,3,#,4,5,6,7,#]
 * <strong>Explanation: </strong>Given the above perfect binary tree (Figure A), your function should populate each next pointer to point to its next right node, just like in Figure B. The serialized output is in level order as connected by the next pointers, with '#' signifying the end of each level.
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 *
 * <pre><strong>Input:</strong> root = []
 * <strong>Output:</strong> []
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li>The number of nodes in the tree is in the range <code>[0, 2<sup>12</sup> - 1]</code>.</li>
 * 	<li><code>-1000 &lt;= Node.val &lt;= 1000</code></li>
 * </ul>
 *
 * <p>&nbsp;</p>
 * <p><strong>Follow-up:</strong></p>
 *
 * <ul>
 * 	<li>You may only use constant extra space.</li>
 * 	<li>The recursive approach is fine. You may assume implicit stack space does not count as extra space for this problem.</li>
 * </ul>
 */
public class PopulatingNextRightPointersInEachNode {
    public static class Node {
        public int val;
        public Node left;
        public Node right;
        public Node next;

        public Node() {
        }

        public Node(int _val) {
            val = _val;
        }

        public Node(int _val, Node _left, Node _right, Node _next) {
            val = _val;
            left = _left;
            right = _right;
            next = _next;
        }
    }

    public static class Solution1 {
        public Node connect(Node root) {
            if (root == null) return null;
            Node dummy = new Node(), tail;
            Node node = root;
            while (node.left != null) {
                tail = dummy;
                while (node != null) {
                    tail = tail.next = node.left;
                    tail = tail.next = node.right;
                    node = node.next;
                }
                node = dummy.next;
            }
            return root;
        }
    }

    public static class Solution2 {
        public Node connect(Node root) {
            if (root == null || root.left == null)
                return root;
            root.left.next = root.right;
            if (root.next != null)
                root.right.next = root.next.left;
            connect(root.left);
            connect(root.right);
            return root;
        }
    }
}
